TestServerStatsServerStreamRPCError failed

```
--- FAIL: TestServerStatsServerStreamRPCError (0.01s)
	stats_test.go:710: - 0, *stats.ConnBegin = &{Client:false}, ctx: context.Background.WithCancel.WithValue(stats_test.connCtxKey{}, &stats.ConnTagInfo{RemoteAddr:(*net.TCPAddr)(0xc820019f50), LocalAddr:(*net.TCPAddr)(0xc8200197d0)})
	stats_test.go:710: - 1, *stats.ConnEnd = &{Client:false}, ctx: context.Background.WithCancel.WithValue(stats_test.connCtxKey{}, &stats.ConnTagInfo{RemoteAddr:(*net.TCPAddr)(0xc820019f50), LocalAddr:(*net.TCPAddr)(0xc8200197d0)})
	stats_test.go:710: - 2, *stats.ConnBegin = &{Client:false}, ctx: context.Background.WithCancel.WithValue(stats_test.connCtxKey{}, &stats.ConnTagInfo{RemoteAddr:(*net.TCPAddr)(0xc8201e24b0), LocalAddr:(*net.TCPAddr)(0xc8201e2480)})
	stats_test.go:710: - 3, *stats.ConnBegin = &{Client:false}, ctx: context.Background.WithCancel.WithValue(stats_test.connCtxKey{}, &stats.ConnTagInfo{RemoteAddr:(*net.TCPAddr)(0xc8205f02a0), LocalAddr:(*net.TCPAddr)(0xc8205f0270)})
	stats_test.go:710: - 4, *stats.ConnEnd = &{Client:false}, ctx: context.Background.WithCancel.WithValue(stats_test.connCtxKey{}, &stats.ConnTagInfo{RemoteAddr:(*net.TCPAddr)(0xc8205f02a0), LocalAddr:(*net.TCPAddr)(0xc8205f0270)})
	stats_test.go:712: got 5 stats, want even positive number
```

--- a/stats/stats_test.go
+++ b/stats/stats_test.go
@@ -1,3 +1,5 @@
+// +build go1.7
+
 /*
  *
  * Copyright 2016 gRPC authors.

--- a/transport/http2_server.go
+++ b/transport/http2_server.go
@@ -228,24 +228,26 @@ func newHTTP2Server(conn net.Conn, config *ServerConfig) (_ ServerTransport, err
 	}
 	t.framer.writer.Flush()
 
+	defer func() {
+		if err != nil {
+			t.Close()
+		}
+	}()
+
 	// Check the validity of client preface.
 	preface := make([]byte, len(clientPreface))
 	if _, err := io.ReadFull(t.conn, preface); err != nil {
-		t.Close()
 		return nil, connectionErrorf(false, err, "transport: http2Server.HandleStreams failed to receive the preface from client: %v", err)
 	}
 	if !bytes.Equal(preface, clientPreface) {
-		t.Close()
 		return nil, connectionErrorf(false, nil, "transport: http2Server.HandleStreams received bogus greeting from client: %q", preface)
 	}
 
 	frame, err := t.framer.fr.ReadFrame()
 	if err == io.EOF || err == io.ErrUnexpectedEOF {
-		t.Close()
 		return nil, err
 	}
 	if err != nil {
-		t.Close()
 		return nil, connectionErrorf(false, err, "transport: http2Server.HandleStreams failed to read initial settings frame: %v", err)
 	}
 	atomic.StoreUint32(&t.activity, 1)

--- a/transport/http2_server.go
+++ b/transport/http2_server.go
@@ -231,18 +231,21 @@ func newHTTP2Server(conn net.Conn, config *ServerConfig) (_ ServerTransport, err
 	// Check the validity of client preface.
 	preface := make([]byte, len(clientPreface))
 	if _, err := io.ReadFull(t.conn, preface); err != nil {
+		t.Close()
 		return nil, connectionErrorf(false, err, "transport: http2Server.HandleStreams failed to receive the preface from client: %v", err)
 	}
 	if !bytes.Equal(preface, clientPreface) {
+		t.Close()
 		return nil, connectionErrorf(false, nil, "transport: http2Server.HandleStreams received bogus greeting from client: %q", preface)
 	}
 
 	frame, err := t.framer.fr.ReadFrame()
 	if err == io.EOF || err == io.ErrUnexpectedEOF {
 		t.Close()
-		return
+		return nil, err
 	}
 	if err != nil {
+		t.Close()
 		return nil, connectionErrorf(false, err, "transport: http2Server.HandleStreams failed to read initial settings frame: %v", err)
 	}
 	atomic.StoreUint32(&t.activity, 1)

